File: frontend/src/api/task.ts
import http from './index'

export interface CreateTaskRequest {
  task_type: string
  urls: string[]
  options?: Record<string, unknown>
}

export interface CreateTaskResponse {
  task_id: string
  status: string
}

export interface TestTask {
  id: string
  user_id: string
  task_type: string
  status: string
  config: string
  progress: number | null
  result: string | null
  error_msg: string | null
  created_at: string
  started_at: string | null
  finished_at: string | null
}

export interface WebsiteResult {
  id: string
  task_id: string
  url: string
  dns_time_ms: number | null
  dns_success: number | null
  tcp_time_ms: number | null
  tls_time_ms: number | null
  http_status: number | null
  ttfb_ms: number | null
  fp_ms: number | null
  fcp_ms: number | null
  dom_content_loaded_ms: number | null
  load_event_ms: number | null
  page_open_time_ms: number | null
  first_paint_ms: number | null
  resource_count: number | null
  resource_total_size: number | null
  final_url: string | null
  page_title: string | null
  screenshot_path: string | null
  error_msg: string | null
  created_at: string
}

export interface VideoResult {
  id: string
  task_id: string
  url: string
  platform: string | null
  first_play_time_ms: number | null
  buffer_count: number | null
  total_buffer_time_ms: number | null
  play_success: number | null
  video_download_speed: number | null
  video_size: number | null
  video_duration_ms: number | null
  dropped_frames: number | null
  decoded_frames: number | null
  screenshot_path: string | null
  page_title: string | null
  error_msg: string | null
  created_at: string
}

export interface DownloadResult {
  id: string
  task_id: string
  url: string
  download_speed: number | null
  avg_speed: number | null
  peak_speed: number | null
  download_time_ms: number | null
  file_size: number | null
  success: number | null
  error_msg: string | null
  created_at: string
}

export interface TaskListResponse {
  tasks: TestTask[]
  total: number
  page: number
  size: number
}

export interface DashboardStats {
  today_tests: number
  success_rate: number
  avg_dns: number
  avg_ttfb: number
  avg_page_time: number
  recent_tasks: TestTask[]
  trend_data: TrendPoint[]
}

export interface TrendPoint {
  time: string
  dns_ms: number
  ttfb_ms: number
  page_ms: number
}

export const taskApi = {
  create(data: CreateTaskRequest) {
    return http.post<unknown, { code: number; msg: string; data: CreateTaskResponse }>(
      '/task/create',
      data,
    )
  },

  list(page = 1, size = 20) {
    return http.get<unknown, { code: number; msg: string; data: TaskListResponse }>(
      '/task/list',
      { params: { page, size } },
    )
  },

  get(taskId: string) {
    return http.get<unknown, { code: number; msg: string; data: TestTask }>(
      `/task/${taskId}`,
    )
  },

  getResults(taskId: string) {
    return http.get<unknown, { code: number; msg: string; data: WebsiteResult[] }>(
      `/task/${taskId}/result`,
    )
  },

  getVideoResults(taskId: string) {
    return http.get<unknown, { code: number; msg: string; data: VideoResult[] }>(
      `/task/${taskId}/video-result`,
    )
  },

  getDownloadResults(taskId: string) {
    return http.get<unknown, { code: number; msg: string; data: DownloadResult[] }>(
      `/task/${taskId}/download-result`,
    )
  },

  cancel(taskId: string) {
    return http.post<unknown, { code: number; msg: string; data: null }>(
      `/task/${taskId}/cancel`,
    )
  },

  retry(taskId: string) {
    return http.post<unknown, { code: number; msg: string; data: CreateTaskResponse }>(
      `/task/${taskId}/retry`,
    )
  },
}

export const dashboardApi = {
  getStats() {
    return http.get<unknown, { code: number; msg: string; data: DashboardStats }>(
      '/dashboard/stats',
    )
  },
}

File: backend/src/models/task.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// 测试任务模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct TestTask {
    pub id: String,
    pub user_id: String,
    pub task_type: String,
    pub status: String,
    pub config: String,
    pub progress: Option<f64>,
    pub result: Option<String>,
    pub error_msg: Option<String>,
    pub created_at: String,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
}

/// 任务类型枚举
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TaskType {
    Website,
    Download,
    Video,
}

impl TaskType {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskType::Website => "website",
            TaskType::Download => "download",
            TaskType::Video => "video",
        }
    }
}

/// 任务状态枚举
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum TaskStatus {
    Pending,
    Running,
    Completed,
    Failed,
    Cancelled,
}

impl TaskStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            TaskStatus::Pending => "pending",
            TaskStatus::Running => "running",
            TaskStatus::Completed => "completed",
            TaskStatus::Failed => "failed",
            TaskStatus::Cancelled => "cancelled",
        }
    }
}

/// 创建任务请求
#[derive(Debug, Deserialize)]
pub struct CreateTaskRequest {
    pub task_type: String,
    pub urls: Vec<String>,
    #[serde(default)]
    pub options: serde_json::Value,
}

/// 创建任务响应
#[derive(Debug, Serialize)]
pub struct CreateTaskResponse {
    pub task_id: String,
    pub status: String,
}

/// 测试配置（从 options 解析，含重复次数）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestConfig {
    #[serde(default = "default_repeat_count")]
    pub repeat_count: usize,
}
fn default_repeat_count() -> usize { 1 }
impl Default for TestConfig {
    fn default() -> Self { Self { repeat_count: 1 } }
}

/// 网站测试结果模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct WebsiteResult {
    pub id: String,
    pub task_id: String,
    pub url: String,
    pub dns_time_ms: Option<f64>,
    pub dns_success: Option<i32>,
    pub tcp_time_ms: Option<f64>,
    pub tls_time_ms: Option<f64>,
    pub http_status: Option<i32>,
    pub ttfb_ms: Option<f64>,
    pub fp_ms: Option<f64>,
    pub fcp_ms: Option<f64>,
    pub dom_content_loaded_ms: Option<f64>,
    pub load_event_ms: Option<f64>,
    pub page_open_time_ms: Option<f64>,
    pub first_paint_ms: Option<f64>,
    pub resource_count: Option<i32>,
    pub resource_total_size: Option<i32>,
    pub html_size: Option<i32>,
    pub css_size: Option<i32>,
    pub js_size: Option<i32>,
    pub image_size: Option<i32>,
    pub font_size: Option<i32>,
    pub total_requests: Option<i32>,
    pub failed_requests: Option<i32>,
    pub lcp_ms: Option<f64>,
    pub cls: Option<f64>,
    pub tti_ms: Option<f64>,
    pub site_size_kb: Option<f64>,
    pub avg_speed_kbps: Option<f64>,
    pub total_speed_kbps: Option<f64>,
    pub first_screen_ratio: Option<f64>,
    pub final_url: Option<String>,
    pub page_title: Option<String>,
    pub screenshot_path: Option<String>,
    pub error_msg: Option<String>,
    pub test_count: Option<i32>,
    pub created_at: String,
}

/// 下载测试结果模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct DownloadResult {
    pub id: String,
    pub task_id: String,
    pub url: String,
    pub dns_time_ms: Option<f64>,
    pub dns_success: Option<i32>,
    pub tcp_time_ms: Option<f64>,
    pub download_speed: Option<f64>,
    pub avg_speed: Option<f64>,
    pub peak_speed: Option<f64>,
    pub download_time_ms: Option<f64>,
    pub file_size: Option<i32>,
    pub success: Option<i32>,
    pub error_msg: Option<String>,
    pub test_count: Option<i32>,
    pub created_at: String,
}

/// 视频测试结果模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct VideoResult {
    pub id: String,
    pub task_id: String,
    pub url: String,
    pub platform: Option<String>,
    pub dns_time_ms: Option<f64>,
    pub dns_success: Option<i32>,
    pub tcp_time_ms: Option<f64>,
    pub http_response_ms: Option<f64>,
    pub first_play_time_ms: Option<f64>,
    pub buffer_count: Option<i32>,
    pub total_buffer_time_ms: Option<f64>,
    pub buffer_rate: Option<f64>,
    pub play_success: Option<i32>,
    pub video_download_speed: Option<f64>,
    pub video_size: Option<i32>,
    pub video_duration_ms: Option<f64>,
    pub dropped_frames: Option<i32>,
    pub decoded_frames: Option<i32>,
    pub screenshot_path: Option<String>,
    pub page_title: Option<String>,
    pub error_msg: Option<String>,
    pub test_count: Option<i32>,
    pub trigger_method: Option<String>,
    pub stutter_count: Option<i32>,
    pub stutter_duration_ms: Option<f64>,
    pub play_duration_sec: Option<f64>,
    pub stutter_ratio: Option<f64>,
    pub video_width: Option<i32>,
    pub video_height: Option<i32>,
    pub video_duration_sec: Option<f64>,
    pub created_at: String,
}

/// 任务日志模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct TaskLog {
    pub id: String,
    pub task_id: String,
    pub level: String,
    pub message: String,
    pub created_at: String,
}

/// Ping 测试结果模型
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PingResult {
    pub id: String,
    pub task_id: String,
    pub host: String,
    pub avg_latency_ms: Option<f64>,
    pub packet_loss_rate: Option<f64>,
    pub jitter_ms: Option<f64>,
    pub success: Option<i32>,
    pub error_msg: Option<String>,
    pub method: Option<String>,
    pub test_count: Option<i32>,
    pub created_at: String,
}
